keycloak/openid_client.go

```diff
@@ -1,37 +1,329 @@
+	if err != nil {
+		return nil, err
+	}
+
+	err = keycloakClient.get(fmt.Sprintf("/realms/%s/clients/%s/authz/resource-server/policy/%s/associatedPolicies", realm, resourceServerId, id), &policies, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	err = keycloakClient.get(fmt.Sprintf("/realms/%s/clients/%s/authz/resource-server/permission/%s/resources", realm, resourceServerId, id), &resources, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, policy := range policies {
+		permission.Policies = append(permission.Policies, policy.Id)
+	}
+
+	for _, resource := range resources {
+		permission.Resources = append(permission.Resources, resource.Id)
+	}
+
+	return &permission, nil
+}
+
+func (keycloakClient *KeycloakClient) NewOpenidClientAuthorizationPermission(permission *OpenidClientAuthorizationPermission) error {
+	body, _, err := keycloakClient.post(fmt.Sprintf("/realms/%s/clients/%s/authz/resource-server/permission", permission.RealmId, permission.ResourceServerId), permission)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(body, &permission)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (keycloakClient *KeycloakClient) UpdateOpenidClientAuthorizationPermission(permission *OpenidClientAuthorizationPermission) error {
+	err := keycloakClient.put(fmt.Sprintf("/realms/%s/clients/%s/authz/resource-server/permission/resource/%s", permission.RealmId, permission.ResourceServerId, permission.Id), permission)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (keycloakClient *KeycloakClient) DeleteOpenidClientAuthorizationPermission(realmId, resourceServerId, permissionId string) error {
+	return keycloakClient.delete(fmt.Sprintf("/realms/%s/clients/%s/authz/resource-server/permission/%s", realmId, resourceServerId, permissionId), nil)
+}
+
+func (keycloakClient *KeycloakClient) GetClientRoleByName(realm, clientId, name string) (*OpenidClientRole, error) {
+	var clientRole OpenidClientRole
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/clients/%s/roles/%s", realm, clientId, name), &clientRole, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &clientRole, nil
+}
+
+func (keycloakClient *KeycloakClient) GetClientByName(realm, clientId string) (*OpenidClient, error) {
+	var clients []OpenidClient
+	params := map[string]string{"clientId": clientId}
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/clients", realm), &clients, params)
+	if err != nil {
+		return nil, err
+	}
+	return &clients[0], nil
+}
+
+func (keycloakClient *KeycloakClient) NewOpenidClientServiceAccountRole(serviceAccountRole *OpenidClientServiceAccountRole) error {
+	role, err := keycloakClient.GetClientRoleByName(serviceAccountRole.RealmId, serviceAccountRole.ContainerId, serviceAccountRole.Name)
+	if err != nil {
+		return err
+	}
+	serviceAccountRole.Id = role.Id
+	serviceAccountRoles := []OpenidClientServiceAccountRole{}
+	serviceAccountRoles = append(serviceAccountRoles, *serviceAccountRole)
```

Comment:
Do you need to use `append` here?  I think you could just initialize the slice with the single element instead of making it empty and adding to it.

Reply:
Done